README.md
## Discription 
The is a comic translate helper tool, uploaded by Isaiah Yang.
(I would rather use the word "uploaded" because 99% of the code are written by other open sources, I just changes a little of it, contains 99.9 % Gimp 2.8.20)
This program is intend to make the translation of comic easier or even automatic. But please don't put it into illegal usage.
Still, it is not perfect and require additional functions to achieve that, however, I would keep working on perfecting it.


## The software are used and needed to be install:

1. Ubuntu (operation system, other linux system not tested yet)
2. Opencv
3. python3.5
	(or any version of python that can install pytesseract and googletrans)
4. Gimp 2.8.20 supplements


## How to install

The fellowing code are executed in ubuntu system.


>  1. step : clone or download the source code
make sure that all your path are in english
my path is only "/helper/"

		git clone https://github.com/kangaroo02/comic_translate_with_gimp.git


>  2. step : install the dependence of gimp
https://ubuntuforums.org/showthread.php?t=1952615&page=3&p=11821399#post11821399


		sudo apt-get update
		sudo apt-get install cvs subversion git-core mercurial
		sudo apt-get build-dep gimp

(If exist error: E: You must put some 'source' URIs in your sources.list
please take a look in this web to open the sources option:
https://askubuntu.com/questions/496549/error-you-must-put-some-source-uris-in-your-sources-list)

		sudo apt-get remove libgegl-0.0-dev libbabl-0.0-0-dev
		sudo apt-get install libavformat-dev libavcodec-dev graphviz liblua5.1-0-dev

(install vala : https://askubuntu.com/questions/910443/how-to-install-vala-on-ubuntu-16-10)

		sudo add-apt-repository ppa:vala-team
		sudo apt update
		sudo apt install valac

(install ruby : https://askubuntu.com/questions/1093390/ubuntu-18-04-how-to-install-ruby1-9-1ruby1-9-1-full-rubygems1-9-1-ruby1-9-1)

		sudo apt install ruby2.5 ruby2.5-dev
		sudo apt-get install libopenexr-dev libopenraw-dev libexiv2-dev \
		libsdl1.2-dev libgs-dev
(install libjasper : https://researchxuyc.wordpress.com/2018/09/26/install-libjasper-in-ubuntu-18-04/)

		sudo add-apt-repository "deb http://security.ubuntu.com/ubuntu xenial-security main"
		sudo apt update
		sudo apt install libjasper1 libjasper-dev

		sudo apt-get install checkinstall


The following three are only valid for that session if you for any reason restart the terminal, you'll have to execute them again

	export PATH=/opt/gimp-2.8/bin:$PATH
	export PKG_CONFIG_PATH=/opt/gimp-2.8/lib/pkgconfig${PKG_CONFIG_PATH:+:${PKG_CONFIG_PATH}}
	export LD_LIBRARY_PATH=/opt/gimp-2.8/lib${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}


(Remember the above three have to be executed again if you restart your session. )




>  3. step : install babl

		cd into the dir of "comic-translate-helper"
		cd babl-0.1.10/
		./configure --prefix=/opt/gimp-2.8
		make
		sudo make install
		cd ..

>  4. step : install gegl

   		cd gegl-0.2.0/
		sudo apt-get install intltool
		./configure --prefix=/opt/gimp-2.8
  	  	make
 	  	sudo make install
 	  	cd ..


>  5. step : install gimp (the comic translate helper)

   		cd gimp-2.8.20/
	

### Install gimp
		sudo apt install python-gtk2 python-gtk2-dev
		

		./configure --prefix=/opt/gimp-2.8
		make
		(this process will take a long time)
		sudo make install
	
		sudo apt install libcanberra-gtk-module libcanberra-gtk3-module


>  6. step : install opencv
https://docs.opencv.org/2.4.13.7/doc/tutorials/introduction/linux_install/linux_install.html
https://pydeeplearning.com/opencv/install-opencv-with-c-on-ubuntu-18-04/

		sudo apt-get update
		sudo apt-get upgrade
		sudo apt-get install libopencv-dev
		sudo pip3 install opencv-contrib-python

>  7. step : install  pytesseract
https://pypi.org/project/pytesseract/
https://github.com/tesseract-ocr/tesseract/wiki

		sudo apt-get install python3-pip
		sudo pip3 install pytesseract
		sudo apt-get install tesseract-ocr
		sudo apt install libtesseract-dev


(download the language package to translate)
japanese

   		sudo apt-get install tesseract-ocr-jpn
english

 	  	sudo apt-get install tesseract-ocr-eng
chinese traditional

  	 	sudo apt-get install tesseract-ocr-chi-tra
	

>  8. step : install googletrans
https://py-googletrans.readthedocs.io/en/latest/

		sudo pip3 install googletrans


>  9. step : rebuild something

		cd to your dir comic-translate-helper
		cd gimp-2.8.20/app/tools/helpercore
		make clear
		make


## Run comic translate helper (actually gimp) in terminal

	export PATH=/opt/gimp-2.8/bin:$PATH
	export PKG_CONFIG_PATH=/opt/gimp-2.8/lib/pkgconfig${PKG_CONFIG_PATH:+:${PKG_CONFIG_PATH}}
	export LD_LIBRARY_PATH=/opt/gimp-2.8/lib${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}
	
	//cd to your gimp2.8 folder
	gimp


>  step : change translate language

	cd gimp-2.8.20/app/tools/helpercore
	edit Detect_and_Translate.py

input:tesseract-ocr language list
https://askubuntu.com/questions/793634/how-do-i-install-a-new-language-pack-for-tesseract-on-16-04

output:googletrans language list
https://py-googletrans.readthedocs.io/en/latest/


gegl-0.2.0/ChangeLog
commit 57e23d310291a632ea8e16b7b8e5b97f7f289680
Author: Øyvind Kolås <pippin@gimp.org>
Date:   Mon Apr 2 22:57:02 2012 +0100

    bump version to 0.2.0

 configure.ac |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

commit 0f519e3d6701718166433df852fd1bcb557cab15
Author: Øyvind Kolås <pippin@gimp.org>
Date:   Mon Apr 2 22:55:50 2012 +0100

    docs: add information about opencl

 docs/index-static.txt.in |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

commit 3231b457726fe5060555b75ccd71202679e138f9
Author: Øyvind Kolås <pippin@gimp.org>
Date:   Mon Apr 2 22:41:10 2012 +0100

    bump babl dependency to 0.1.10

 configure.ac |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit dc0bf2c5e9e25ab0b0b9d344942e52fd2fdf170e
Author: Øyvind Kolås <pippin@gimp.org>
Date:   Mon Apr 2 22:40:55 2012 +0100

    mark gamma test as broken

 tests/compositions/Makefile.am                |    2 +-
 tests/compositions/gamma-broken.xml           |   13 +++++++++++++
 tests/compositions/gamma.xml                  |   13 -------------
 tests/compositions/reference/gamma-broken.png |  Bin 0 -> 161310 bytes
 tests/compositions/reference/gamma.png        |  Bin 161310 -> 0 bytes
 5 files changed, 14 insertions(+), 14 deletions(-)

commit 376e3fef0cd33039f710508f67bded202f542583
Author: Jan Rüegg <rggjan@gmail.com>
Date:   Mon Mar 26 21:38:43 2012 +0200

    global-matting: resynced with rggjan's repo

    ! Changed matting reference image
    ! Fixed Copyright
    + Added levels to gegl chant for matting
    + Better adaption to gimp HACKORDNUNG
    c Adapted user message
    c Cleanup
    - Removed unused function
    ! Fixed nasty memory bug

 operations/common/matting-global.c              |  542
 ++++++++++-------------
 tests/compositions/matting-global.xml           |    7 +-
 tests/compositions/reference/matting-global.png |  Bin 52292 ->
 1241155 bytes
 3 files changed, 236 insertions(+), 313 deletions(-)

commit 5ce5671d35b3b5c0cfcb05e847228001b40cdbee
Author: Mikael Magnusson <mikachu@src.gnome.org>
Date:   Mon Apr 2 21:42:50 2012 +0200

    NEWS: remove doubled spaces

 NEWS |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

commit 163625cfe0b390a59bc8afeb18d144f2be9040a3
Author: Øyvind Kolås <pippin@gimp.org>
Date:   Mon Apr 2 18:51:43 2012 +0100

    gegl-init: default to not use OpenCL

    For the upcoming release only,. it is not stable enough to be
    enabled by
    defualt, memory leaks and some imprecisions are currently blocking it.

 gegl/gegl-init.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 68ccdb92b1e1f8b16206dd17ce5e6598dccec6eb
Author: Øyvind Kolås <pippin@gimp.org>
Date:   Mon Apr 2 18:50:29 2012 +0100

    use non-cl rendering for clones test

 tests/compositions/reference/clones.png |  Bin 766734 -> 789205 bytes
 1 files changed, 0 insertions(+), 0 deletions(-)

commit a937188c2cb49e013c5dd39fdaa5c5b8743a85c8
Author: Øyvind Kolås <pippin@gimp.org>
Date:   Mon Apr 2 18:31:42 2012 +0100

    operation: remove processor code for SIMD dispatch

    No longer needed, using OpenCL is better.

 gegl/operation/Makefile.am                 |    1 -
 gegl/operation/gegl-operation-processors.c |  246
 ----------------------------
 gegl/operation/gegl-operation.h            |    6 -
 3 files changed, 0 insertions(+), 253 deletions(-)

commit 6a0e67887ce98c234022414c5c8a69305ada9fe8
Author: Øyvind Kolås <pippin@gimp.org>
Date:   Mon Apr 2 18:21:23 2012 +0100

    NEWS: add Jon Nordby to contributors

 NEWS |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

commit 8671f62be72429231cc840a415da366daf105393
Author: Øyvind Kolås <pippin@gimp.org>
Date:   Mon Apr 2 18:20:01 2012 +0100

    update reference image for clones, it is correct again

 tests/compositions/reference/clones.png |  Bin 789206 -> 766734 bytes
 1 files changed, 0 insertions(+), 0 deletions(-)

commit a5db476afebd8d6db4ef61298f2e7ff20b9fa797
Author: Massimo Valentini <mvalentini@src.gnome.org>
Date:   Mon Apr 2 19:03:38 2012 +0200

    Bug 656272: Affine transformations are wrong

 operations/transform/transform-core.c              |    4 ++--
 tests/compositions/reference/clones.png            |  Bin 802439 ->
 789206 bytes
 .../compositions/reference/composite-transform.png |  Bin 16350 ->
 16378 bytes
 3 files changed, 2 insertions(+), 2 deletions(-)

commit facbddda49c43596b694b4098f4a97cdd74684c5
Author: Jon Nordby <jononor@gmail.com>
Date:   Mon Apr 2 18:55:29 2012 +0200

    Remove gegl_processor_destroy also from header

    The function was removed in
    a6c6b5fd99f33f04608b05cd0930a22bc3f8603e

 gegl/gegl.h                   |    2 +-
 gegl/process/gegl-processor.h |    2 --
 2 files changed, 1 insertions(+), 3 deletions(-)

commit 01abcf617753df34e149a9633278c1be26b74231
Author: Øyvind Kolås <pippin@gimp.org>
Date:   Mon Apr 2 17:04:43 2012 +0100

    buffer: remove remnants of simd data type from cubic sampler

 gegl/buffer/gegl-sampler-cubic.c |   20 --------------------
 1 files changed, 0 insertions(+), 20 deletions(-)

commit 8c4f835f2f617e150a8ee0bbe53de54243e2dc54
Author: Øyvind Kolås <pippin@gimp.org>
Date:   Mon Apr 2 16:15:16 2012 +0100

    GeglColor: remove gegl_color_get_rgba4f

    gegl_color_get_pixel with babl_format ("RGBA float"), serves the
    same purpose.

 gegl/property-types/gegl-color.c      |   11 -----------
 gegl/property-types/gegl-color.h      |   11 -----------
 operations/common/checkerboard.c      |    4 ++--
 operations/common/color-to-alpha.c    |    2 +-
 operations/common/color.c             |    2 +-
 operations/common/grid.c              |    2 +-
 operations/common/vignette.c          |    4 ++--
 operations/external/path.c            |    2 +-
 operations/external/text.c            |    2 +-
 operations/workshop/linear-gradient.c |    4 ++--
 operations/workshop/radial-gradient.c |    4 ++--
 11 files changed, 13 insertions(+), 35 deletions(-)


gimp-2.8.20/app/core/gimpmarshal.h

#ifndef __gimp_marshal_MARSHAL_H__
#define __gimp_marshal_MARSHAL_H__

#include	<glib-object.h>

G_BEGIN_DECLS

/* BOOLEAN:BOOLEAN (./gimpmarshal.list:25) */
extern void gimp_marshal_BOOLEAN__BOOLEAN (GClosure     *closure,
                                           GValue       *return_value,
                                           guint         n_param_values,
                                           const GValue *param_values,
                                           gpointer      invocation_hint,
                                           gpointer      marshal_data);

/* BOOLEAN:ENUM,INT (./gimpmarshal.list:26) */
extern void gimp_marshal_BOOLEAN__ENUM_INT (GClosure     *closure,
                                            GValue       *return_value,
                                            guint         n_param_values,
                                            const GValue *param_values,
                                            gpointer      invocation_hint,
                                            gpointer      marshal_data);

/* BOOLEAN:OBJECT,POINTER (./gimpmarshal.list:27) */
extern void gimp_marshal_BOOLEAN__OBJECT_POINTER (GClosure     *closure,
                                                  GValue       *return_value,
                                                  guint         n_param_values,
                                                  const GValue *param_values,
                                                  gpointer      invocation_hint,
                                                  gpointer      marshal_data);

/* BOOLEAN:OBJECT,POINTER,STRING (./gimpmarshal.list:28) */
extern void gimp_marshal_BOOLEAN__OBJECT_POINTER_STRING (GClosure     *closure,
                                                         GValue       *return_value,
                                                         guint         n_param_values,
                                                         const GValue *param_values,
                                                         gpointer      invocation_hint,
                                                         gpointer      marshal_data);

/* BOOLEAN:STRING (./gimpmarshal.list:29) */
extern void gimp_marshal_BOOLEAN__STRING (GClosure     *closure,
                                          GValue       *return_value,
                                          guint         n_param_values,
                                          const GValue *param_values,
                                          gpointer      invocation_hint,
                                          gpointer      marshal_data);

/* BOOLEAN:STRING,FLAGS (./gimpmarshal.list:30) */
extern void gimp_marshal_BOOLEAN__STRING_FLAGS (GClosure     *closure,
                                                GValue       *return_value,
                                                guint         n_param_values,
                                                const GValue *param_values,
                                                gpointer      invocation_hint,
                                                gpointer      marshal_data);

/* VOID:BOOLEAN (./gimpmarshal.list:32) */
#define gimp_marshal_VOID__BOOLEAN	g_cclosure_marshal_VOID__BOOLEAN

/* VOID:BOOLEAN,INT,INT,INT,INT (./gimpmarshal.list:33) */
extern void gimp_marshal_VOID__BOOLEAN_INT_INT_INT_INT (GClosure     *closure,
                                                        GValue       *return_value,
                                                        guint         n_param_values,
                                                        const GValue *param_values,
                                                        gpointer      invocation_hint,
                                                        gpointer      marshal_data);

/* VOID:BOXED (./gimpmarshal.list:34) */
#define gimp_marshal_VOID__BOXED	g_cclosure_marshal_VOID__BOXED

/* VOID:BOXED,ENUM (./gimpmarshal.list:35) */
extern void gimp_marshal_VOID__BOXED_ENUM (GClosure     *closure,
                                           GValue       *return_value,
                                           guint         n_param_values,
                                           const GValue *param_values,
                                           gpointer      invocation_hint,
                                           gpointer      marshal_data);

/* VOID:DOUBLE (./gimpmarshal.list:36) */
#define gimp_marshal_VOID__DOUBLE	g_cclosure_marshal_VOID__DOUBLE

/* VOID:DOUBLE,DOUBLE (./gimpmarshal.list:37) */
extern void gimp_marshal_VOID__DOUBLE_DOUBLE (GClosure     *closure,
                                              GValue       *return_value,
                                              guint         n_param_values,
                                              const GValue *param_values,
                                              gpointer      invocation_hint,
                                              gpointer      marshal_data);

/* VOID:DOUBLE,DOUBLE,DOUBLE,DOUBLE (./gimpmarshal.list:38) */
extern void gimp_marshal_VOID__DOUBLE_DOUBLE_DOUBLE_DOUBLE (GClosure     *closure,
                                                            GValue       *return_value,
                                                            guint         n_param_values,
                                                            const GValue *param_values,
                                                            gpointer      invocation_hint,
                                                            gpointer      marshal_data);

/* VOID:ENUM (./gimpmarshal.list:39) */
#define gimp_marshal_VOID__ENUM	g_cclosure_marshal_VOID__ENUM

/* VOID:ENUM,ENUM,BOXED,INT (./gimpmarshal.list:40) */
extern void gimp_marshal_VOID__ENUM_ENUM_BOXED_INT (GClosure     *closure,
                                                    GValue       *return_value,
                                                    guint         n_param_values,
                                                    const GValue *param_values,
                                                    gpointer      invocation_hint,
                                                    gpointer      marshal_data);

/* VOID:ENUM,INT (./gimpmarshal.list:41) */
extern void gimp_marshal_VOID__ENUM_INT (GClosure     *closure,
                                         GValue       *return_value,
                                         guint         n_param_values,
                                         const GValue *param_values,
                                         gpointer      invocation_hint,
                                         gpointer      marshal_data);

/* VOID:ENUM,INT,BOOLEAN (./gimpmarshal.list:42) */
extern void gimp_marshal_VOID__ENUM_INT_BOOLEAN (GClosure     *closure,
                                                 GValue       *return_value,
                                                 guint         n_param_values,
                                                 const GValue *param_values,
                                                 gpointer      invocation_hint,
                                                 gpointer      marshal_data);

/* VOID:ENUM,OBJECT (./gimpmarshal.list:43) */
extern void gimp_marshal_VOID__ENUM_OBJECT (GClosure     *closure,
                                            GValue       *return_value,
                                            guint         n_param_values,
                                            const GValue *param_values,
                                            gpointer      invocation_hint,
                                            gpointer      marshal_data);

/* VOID:ENUM,POINTER (./gimpmarshal.list:44) */
extern void gimp_marshal_VOID__ENUM_POINTER (GClosure     *closure,
                                             GValue       *return_value,
                                             guint         n_param_values,
                                             const GValue *param_values,
                                             gpointer      invocation_hint,
                                             gpointer      marshal_data);

/* VOID:FLAGS (./gimpmarshal.list:45) */
#define gimp_marshal_VOID__FLAGS	g_cclosure_marshal_VOID__FLAGS

/* VOID:INT (./gimpmarshal.list:46) */
#define gimp_marshal_VOID__INT	g_cclosure_marshal_VOID__INT

/* VOID:INT,INT (./gimpmarshal.list:47) */
extern void gimp_marshal_VOID__INT_INT (GClosure     *closure,
                                        GValue       *return_value,
                                        guint         n_param_values,
                                        const GValue *param_values,
                                        gpointer      invocation_hint,
                                        gpointer      marshal_data);

/* VOID:INT,INT,INT,INT (./gimpmarshal.list:48) */
extern void gimp_marshal_VOID__INT_INT_INT_INT (GClosure     *closure,
                                                GValue       *return_value,
                                                guint         n_param_values,
                                                const GValue *param_values,
                                                gpointer      invocation_hint,
                                                gpointer      marshal_data);

/* VOID:OBJECT (./gimpmarshal.list:49) */
#define gimp_marshal_VOID__OBJECT	g_cclosure_marshal_VOID__OBJECT

/* VOID:OBJECT,BOOLEAN (./gimpmarshal.list:50) */
extern void gimp_marshal_VOID__OBJECT_BOOLEAN (GClosure     *closure,
                                               GValue       *return_value,
                                               guint         n_param_values,
                                               const GValue *param_values,
                                               gpointer      invocation_hint,
                                               gpointer      marshal_data);

/* VOID:OBJECT,INT (./gimpmarshal.list:51) */
extern void gimp_marshal_VOID__OBJECT_INT (GClosure     *closure,
                                           GValue       *return_value,
                                           guint         n_param_values,
                                           const GValue *param_values,
                                           gpointer      invocation_hint,
                                           gpointer      marshal_data);

/* VOID:OBJECT,OBJECT (./gimpmarshal.list:52) */
extern void gimp_marshal_VOID__OBJECT_OBJECT (GClosure     *closure,
                                              GValue       *return_value,
                                              guint         n_param_values,
                                              const GValue *param_values,
                                              gpointer      invocation_hint,
                                              gpointer      marshal_data);

/* VOID:OBJECT,POINTER (./gimpmarshal.list:53) */
extern void gimp_marshal_VOID__OBJECT_POINTER (GClosure     *closure,
                                               GValue       *return_value,
                                               guint         n_param_values,
                                               const GValue *param_values,
                                               gpointer      invocation_hint,
                                               gpointer      marshal_data);

/* VOID:POINTER (./gimpmarshal.list:54) */
#define gimp_marshal_VOID__POINTER	g_cclosure_marshal_VOID__POINTER

/* VOID:POINTER,BOXED (./gimpmarshal.list:55) */
extern void gimp_marshal_VOID__POINTER_BOXED (GClosure     *closure,
                                              GValue       *return_value,
                                              guint         n_param_values,
                                              const GValue *param_values,
                                              gpointer      invocation_hint,
                                              gpointer      marshal_data);

/* VOID:POINTER,ENUM (./gimpmarshal.list:56) */
extern void gimp_marshal_VOID__POINTER_ENUM (GClosure     *closure,
                                             GValue       *return_value,
                                             guint         n_param_values,
                                             const GValue *param_values,
                                             gpointer      invocation_hint,
                                             gpointer      marshal_data);

/* VOID:POINTER,FLAGS,BOOLEAN (./gimpmarshal.list:57) */
extern void gimp_marshal_VOID__POINTER_FLAGS_BOOLEAN (GClosure     *closure,
                                                      GValue       *return_value,
                                                      guint         n_param_values,
                                                      const GValue *param_values,
                                                      gpointer      invocation_hint,
                                                      gpointer      marshal_data);

/* VOID:POINTER,UINT,FLAGS (./gimpmarshal.list:58) */
extern void gimp_marshal_VOID__POINTER_UINT_FLAGS (GClosure     *closure,
                                                   GValue       *return_value,
                                                   guint         n_param_values,
                                                   const GValue *param_values,
                                                   gpointer      invocation_hint,
                                                   gpointer      marshal_data);

/* VOID:STRING (./gimpmarshal.list:59) */
#define gimp_marshal_VOID__STRING	g_cclosure_marshal_VOID__STRING

/* VOID:STRING,BOOLEAN,UINT,FLAGS (./gimpmarshal.list:60) */
extern void gimp_marshal_VOID__STRING_BOOLEAN_UINT_FLAGS (GClosure     *closure,
                                                          GValue       *return_value,
                                                          guint         n_param_values,
                                                          const GValue *param_values,
                                                          gpointer      invocation_hint,
                                                          gpointer      marshal_data);

/* VOID:STRING,FLAGS (./gimpmarshal.list:61) */
extern void gimp_marshal_VOID__STRING_FLAGS (GClosure     *closure,
                                             GValue       *return_value,
                                             guint         n_param_values,
                                             const GValue *param_values,
                                             gpointer      invocation_hint,
                                             gpointer      marshal_data);

/* VOID:STRING,STRING,STRING (./gimpmarshal.list:62) */
extern void gimp_marshal_VOID__STRING_STRING_STRING (GClosure     *closure,
                                                     GValue       *return_value,
                                                     guint         n_param_values,
                                                     const GValue *param_values,
                                                     gpointer      invocation_hint,
                                                     gpointer      marshal_data);

/* VOID:VOID (./gimpmarshal.list:63) */
#define gimp_marshal_VOID__VOID	g_cclosure_marshal_VOID__VOID

G_END_DECLS

#endif /* __gimp_marshal_MARSHAL_H__ */

